compiler/noirc_evaluator/src/ssa/opt/inline_const_brillig_calls.rs

```diff
@@ -0,0 +1,221 @@
+                    brillig_functions,
+                    brillig_functions_we_could_not_inline,
+                    inliner_aggressiveness,
+                    error_selector_to_type,
+                ) {
+                    self.dfg[block_id].instructions_mut().push(instruction_id);
+                }
+            }
+        }
+    }
+
+    /// Tries to optimize an instruction if it's a call that points to a brillig function,
+    /// and all its arguments are constant. If the optimization is successful, the
+    /// values returned by the brillig call are replaced by the constant values that the
+    /// function returns, and this method returns `true`.
+    fn optimize_const_brillig_call(
+        &mut self,
+        instruction_id: InstructionId,
+        brillig_functions: &BTreeMap<FunctionId, Function>,
+        brillig_functions_we_could_not_inline: &mut HashSet<FunctionId>,
+        inliner_aggressiveness: i64,
+        error_selector_to_type: &BTreeMap<ErrorSelector, Type>,
+    ) -> bool {
+        let instruction = &self.dfg[instruction_id];
+        let Instruction::Call { func: func_id, arguments } = instruction else {
+            return false;
+        };
+
+        let func_value = &self.dfg[*func_id];
+        let Value::Function(func_id) = func_value else {
+            return false;
+        };
+
+        let Some(function) = brillig_functions.get(func_id) else {
+            return false;
+        };
+
+        if !arguments.iter().all(|argument| self.dfg.is_constant(*argument)) {
+            brillig_functions_we_could_not_inline.insert(*func_id);
+            return false;
+        }
+
+        // The function we have is already a copy of the original function, but we need to clone
+        // it again because there might be multiple calls to the same brillig function.
+        let mut function = Function::clone_with_id(*func_id, function);
+
+        // Find the entry block and remove its parameters
+        let entry_block_id = function.entry_block();
+        let entry_block = &mut function.dfg[entry_block_id];
+        let entry_block_parameters = entry_block.take_parameters();
+
+        assert_eq!(arguments.len(), entry_block_parameters.len());
+
+        // Replace the ValueId of parameters with the ValueId of arguments
+        for (parameter_id, argument_id) in entry_block_parameters.iter().zip(arguments) {
+            // Lookup the argument in the current function and insert it in the function copy
+            let new_argument_id = self.copy_constant_to_function(*argument_id, &mut function);
+            function.dfg.set_value_from_id(*parameter_id, new_argument_id);
+        }
+
+        // Try to fully optimize the function. If we can't, we can't inline it's constant value.
+        let Ok(mut function) = optimize(function, inliner_aggressiveness, error_selector_to_type)
+        else {
+            brillig_functions_we_could_not_inline.insert(*func_id);
+            return false;
```

Comment:
Just because you said you didn't like similar repetition, it could possibly be wrapped up into an enum with variants like `NotBrillig`, `CannotInline(FunctionId)`, `Optimized` and then the caller could make sure to insert into the `brillig_functions_we_could_not_inline` instead of having to pass it in, and also to return the instruction to the block unless it's `Optimized`.

Reply:
This is a great idea! I pushed a commit for this. I think it simplified some things, and it's actually the way I wanted the code to look like, but wasn't sure how to do it.